[LBSE] Enable compositing support for SVG elements

nikolaszimmermann opened the pull request:
#### 5c46074bb656ecdbbc05d4d465e78416fad1ffe8
<pre>
[LBSE] Enable compositing support for SVG elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=242833">https://bugs.webkit.org/show_bug.cgi?id=242833</a>

Reviewed by Simon Fraser.

Enable compositing / 3D-transforms / z-index for SVG.

To make compositing useful in the context of SVG, a long standing bug needs to be fixed:
webkit.org/b/27684 (&quot;Composited elements appear pixelated when scaled up using transform&quot;).
The idea is to include the maximum scale factor a CSS transform can induce into the &quot;content
scale factor&quot; belonging to the backing of the layer. Otherwise CSS transforms have no impact
on the internal GraphicsLayer size --&gt; we&apos;ll render at scale=1 and upscale according to the
CSS transform. This results in blurry rendering. This has been attempted by quit a few people
over the years, however there were reservations:
- JS driven, e.g. rAF() animations, that modify scale e.g. from 1 -&gt; 2 in 0.001 steps. Previously
  we never re-renderered, now we&apos;d to it on every scale change. Needs some heuristics, a treshold,
  (|old-new|/old&gt;x?) to enable this or stay with upscaling behavior.

- For WebAnimations/CSS animations/transitions we&apos;d need to figure out the max scale for the whole
  animation timeline, so that we can figure it out once, and can avoid re-rendering during the animation.
  (This is not implemented in this patch yet, but my colleague Miguel has a patch for that...)

The question is: do we want a setting to toggle this at runtime? Or a CSS prop so that authors can
express their intent? (Please never render blurry, or I don&apos;t care, but want max. performance...).
SVG definately needs sharp graphics without upscaling artifacts: there are documents defined in a
10px x 10px space that are upscaling to window size (e.g. 800px / 600px) that would look completly
blurry with the current upscaling strategy.
Therefore I decided to introduce a setShouldUpdateRootRelativeScaleFactor() setter for GraphicsLayer,
that&apos;s only used by LBSE for now to activate the new strategy, so that I can move on with LBSE
upstreaming. However it would be the right time now to generalize this for all ports and settle
on a future-proof strategy how to deal with this issue. Web Authors already waited way too long
for a solution :-)

3D transformations work out of the box now for every SVG element, tested under various conditions
&lt;svg&gt; embedded in HTML, with CSS box model properties applied, such as border/padding/margin
with/without overflow clipping, with/without non-default transform-origin etc.

To make z-index work, we only have to switch to a new logic to respect z-index for non-positioned
elements (position: static, is the default for SVG elements) in StyleAdjuster, not more.

The bulk of the prepartions were already upstreamed, this smallish patch finally enables the
new fance SVG2+ features.

Add new reftests that cover SVG compositing with 2D/3D transforms (LayoutTests/svg/compositing)
and new z-index reftests (LayoutTests/svg/z-index). These news tests enable LBSE at runtime
such that these features are tested by default on macOS/iOS platforms on EWS. To switch on LBSE
at runtime these tests utilize following marker: &lt;!-- webkit-test-runner [ LayerBasedSVGEngineEnabled=true ] --&gt;

Covered by new and existing tests.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child-overflow-hidden-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child-overflow-hidden.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-shape-child-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-shape-child.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-transformed-group-child-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-transformed-group-child.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-overflow-visible-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-overflow-visible.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-margin-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-margin.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border.html: Added.
* LayoutTests/svg/compositing/svg-poster-circle-expected.html: Added.
* LayoutTests/svg/compositing/svg-poster-circle.html: Added.
* LayoutTests/svg/z-index/group-with-zindex-expected.svg: Added.
* LayoutTests/svg/z-index/group-with-zindex.svg: Added.
* LayoutTests/svg/z-index/shapes-with-zindex-expected.svg: Added.
* LayoutTests/svg/z-index/shapes-with-zindex.svg: Added.
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setShouldUpdateRootRelativeScaleFactor):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::updateRootRelativeScale):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::createGraphicsLayer):
(WebCore::RenderLayerBacking::paintsContent const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
(WebCore::Style::Adjuster::adjustSVGElementStyle):

Canonical link: <a href="https://commits.webkit.org/253054@main">https://commits.webkit.org/253054@main</a>
</pre>

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
```diff
@@ -659,6 +661,8 @@ class GraphicsLayerCA : public GraphicsLayer, public PlatformCALayerClient {
     std::unique_ptr<DisplayList::InMemoryDisplayList> m_displayList;
 
     float m_contentsScaleLimitingFactor { 1 };
+    bool m_shouldUpdateRootRelativeScaleFactor { false };
```
Packing is non-optimal here (you waste 3 bytes).

nikolaszimmermann replied:
Hmm, do we support bool m_foo : 1 { false } already? I'll check if we have that C++20 feature at hand.

nikolaszimmermann replied:
Oh it's already used widely, nice.

```
diff --git a/Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h b/Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h
index 367ad35c845d..a6df8d50c480 100644
--- a/Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h
+++ b/Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h
@@ -661,12 +661,11 @@ private:
     std::unique_ptr<DisplayList::InMemoryDisplayList> m_displayList;
 
     float m_contentsScaleLimitingFactor { 1 };
-    bool m_shouldUpdateRootRelativeScaleFactor { false };
     float m_rootRelativeScaleFactor { 1.0f };
 
     ContentsLayerPurpose m_contentsLayerPurpose { ContentsLayerPurpose::None };
-    bool m_isCommittingChanges { false };
-
+    bool m_isCommittingChanges : 1 { false };
+    bool m_shouldUpdateRootRelativeScaleFactor : 1 { false };
     bool m_needsFullRepaint : 1;
     bool m_allowsBackingStoreDetaching : 1;
     bool m_intersectsCoverageRect : 1;
```

This removes wasting space here and in another place + sorted for minimal padding.

nikolaszimmermann replied:
```
./platform/graphics/ca/GraphicsLayerCA.cpp:1954:41: error: non-const reference cannot bind to bit-field 'm_isCommittingChanges'
    SetForScope committingChangesChange(m_isCommittingChanges, true);
```

Oops, m_isCommittingChanges cannot carry a bit field suffix. Will fix and land.

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2688,6 +2693,17 @@ bool RenderLayerBacking::paintsContent(RenderLayer::PaintedContentRequest& reque
         return paintsContent;
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    // FIXME: [LBSE] Eventually refine the logic to end up with a narrower set of conditions.
```
Please file a bug to track this.

nikolaszimmermann replied:
Done. Added a link to webkit.org/b/243417.

nikolaszimmermann committed:
[LBSE] Enable compositing support for SVG elements
https://bugs.webkit.org/show_bug.cgi?id=242833

Reviewed by Simon Fraser.

Enable compositing / 3D-transforms / z-index for SVG.

To make compositing useful in the context of SVG, a long standing bug needs to be fixed:
webkit.org/b/27684 ("Composited elements appear pixelated when scaled up using transform").
The idea is to include the maximum scale factor a CSS transform can induce into the "content
scale factor" belonging to the backing of the layer. Otherwise CSS transforms have no impact
on the internal GraphicsLayer size --> we'll render at scale=1 and upscale according to the
CSS transform. This results in blurry rendering. This has been attempted by quit a few people
over the years, however there were reservations:
- JS driven, e.g. rAF() animations, that modify scale e.g. from 1 -> 2 in 0.001 steps. Previously
  we never re-renderered, now we'd to it on every scale change. Needs some heuristics, a treshold,
  (|old-new|/old>x?) to enable this or stay with upscaling behavior.

- For WebAnimations/CSS animations/transitions we'd need to figure out the max scale for the whole
  animation timeline, so that we can figure it out once, and can avoid re-rendering during the animation.
  (This is not implemented in this patch yet, but my colleague Miguel has a patch for that...)

The question is: do we want a setting to toggle this at runtime? Or a CSS prop so that authors can
express their intent? (Please never render blurry, or I don't care, but want max. performance...).
SVG definately needs sharp graphics without upscaling artifacts: there are documents defined in a
10px x 10px space that are upscaling to window size (e.g. 800px / 600px) that would look completly
blurry with the current upscaling strategy.
Therefore I decided to introduce a setShouldUpdateRootRelativeScaleFactor() setter for GraphicsLayer,
that's only used by LBSE for now to activate the new strategy, so that I can move on with LBSE
upstreaming. However it would be the right time now to generalize this for all ports and settle
on a future-proof strategy how to deal with this issue. Web Authors already waited way too long
for a solution :-)

3D transformations work out of the box now for every SVG element, tested under various conditions
<svg> embedded in HTML, with CSS box model properties applied, such as border/padding/margin
with/without overflow clipping, with/without non-default transform-origin etc.

To make z-index work, we only have to switch to a new logic to respect z-index for non-positioned
elements (position: static, is the default for SVG elements) in StyleAdjuster, not more.

The bulk of the prepartions were already upstreamed, this smallish patch finally enables the
new fance SVG2+ features.

Add new reftests that cover SVG compositing with 2D/3D transforms (LayoutTests/svg/compositing)
and new z-index reftests (LayoutTests/svg/z-index). These news tests enable LBSE at runtime
such that these features are tested by default on macOS/iOS platforms on EWS. To switch on LBSE
at runtime these tests utilize following marker: <!-- webkit-test-runner [ LayerBasedSVGEngineEnabled=true ] -->

Covered by new and existing tests.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child-overflow-hidden-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child-overflow-hidden.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-shape-child-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-shape-child.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-transformed-group-child-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-directly-composited-transformed-group-child.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-overflow-visible-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-overflow-visible.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-margin-expected.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-margin.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border-padding.html: Added.
* LayoutTests/svg/compositing/outermost-svg-with-border.html: Added.
* LayoutTests/svg/compositing/svg-poster-circle-expected.html: Added.
* LayoutTests/svg/compositing/svg-poster-circle.html: Added.
* LayoutTests/svg/z-index/group-with-zindex-expected.svg: Added.
* LayoutTests/svg/z-index/group-with-zindex.svg: Added.
* LayoutTests/svg/z-index/shapes-with-zindex-expected.svg: Added.
* LayoutTests/svg/z-index/shapes-with-zindex.svg: Added.
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setShouldUpdateRootRelativeScaleFactor):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::updateRootRelativeScale):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::createGraphicsLayer):
(WebCore::RenderLayerBacking::paintsContent const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
(WebCore::Style::Adjuster::adjustSVGElementStyle):

Canonical link: https://commits.webkit.org/253054@main
